### PR TITLE
Fix Meta-Llama-3.1-8B-Instruct chat template

### PR DESCRIPTION
**Pull request by pavel-esir**
In order for Llama-3.1to work need to replace none's with undefined.

Also replaced ValuesList -> GenericMap for slicing callback to work
https://github.com/jinja2cpp/Jinja2Cpp/blob/b32fbde7d98d13c34784c332c4a24a6f92c76e38/test/user_callable_test.cpp#L234

ticket CVS-149027

### CODE REVIEW

File: src/cpp/src/tokenizer.cpp [hunk 1]
```diff
@@ -353,7 +353,8 @@ class Tokenizer::TokenizerImpl {
         std::pair<std::string, std::string> replace_str_map[] = {
             {"'}", "' }"},
             {"{'", "{ '"},
-            {".strip()", ""}
+            {".strip()", ""},
+            {"is none", "is undefined"}
```
Comment (ilya-lavrenov) [inline src/cpp/src/tokenizer.cpp:357]: can we add `Meta-Llama-3.1-8B-Instruct` template to https://github.com/openvinotoolkit/openvino.genai/blob/master/tests/python_tests/tokenizer_configs.py ?

Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp:357]: Good Idea. If it's not a full template fix, you can add only a part the template.

Comment (pavel-esir) [reply src/cpp/src/tokenizer.cpp:357]: Yes, sure. When the fix is ready i will add test case for that

Comment (pavel-esir) [reply src/cpp/src/tokenizer.cpp]: done

File: src/cpp/src/tokenizer.cpp [hunk 2]
```diff
@@ -335,6 +337,31 @@ class Tokenizer::TokenizerImpl {
...
+            {"'}", "' }"},
+            {"{'", "{ '"},
+            {".strip()", ""},
+            {"is not none", "is defined"},
+            {"is none", "is undefined"},
+            {"= none", "= undefined"},
+            // TODO: Jinja2Cpp does not support Python-style slicing, e.g. [1:].
+            // If chat template contains such slicing, we replace it with 
+            // a placeholder at the moment.
+            {"messages[1:]", "slice(messages, 1)"},
```
Comment (Wovchena) [inline src/cpp/src/tokenizer.cpp:352]: What is supposed to be done in the scope of TODO?

File: src/cpp/src/tokenizer.cpp [hunk 3]
```diff
@@ -374,39 +387,28 @@ class Tokenizer::TokenizerImpl {
...
-                int64_t stop = list.size();
-                int64_t step = 1;
-                for (int64_t i = start; i < stop && i < list.size(); i += step)
-                {
-                    result.push_back(list.at(i));
+
+                size_t iter_num = 0;
+                for (auto message = messages.begin(); message != messages.end(); message++, iter_num++) {
+                    if (iter_num < start)
+                        continue;
```
Comment (Wovchena) [inline src/cpp/src/tokenizer.cpp:403]: You can probably initialize `message` iterator with `messages.begin() + start`.

Even better. std::vector allows construction from iterators. Maybe `jinja2::ValuesList` can do that too. If that's the cause it all collapses to `return jinja2::ValuesList{messages.begin() + start, messages.end());`

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: src/cpp/src/tokenizer.cpp (hunk 3 above)
Comment (pavel-esir) [reply src/cpp/src/tokenizer.cpp:403]: >You can probably initialize message iterator with messages.begin() + start.
I tried, it didn't work, i tried also `messages.GetAccessor()` still no luck. So this is the simples solution i came up with

**Review by Wovchena — APPROVED** [review]

File: src/cpp/src/tokenizer.cpp (hunk 2 above)
Comment (pavel-esir) [reply src/cpp/src/tokenizer.cpp:352]: I forget to remove this TODO when i fixed callback. Will remove it

**Review by alexeytarakanov — APPROVED** [review]

File: src/cpp/src/tokenizer.cpp (hunk 3 above)
Comment (Wovchena) [reply src/cpp/src/tokenizer.cpp:403]: Does
>initialize `message` iterator with `messages.begin() + start`

work?

**Pull request merged**